COMP: change target of file system completion for path string literals

Kobzol opened the pull request:
This PR changes the context of filesystem path completion for path-like string literals. They will now resolve to the containing workspace root.

Fixes: https://github.com/intellij-rust/intellij-rust/issues/7219

changelog: Change the completion context of path string literals to the containing workspace root.

Undin commented on src/main/kotlin/org/rust/lang/core/resolve/ref/RsLitExprReferenceContributor.kt:
```diff
@@ -59,7 +59,11 @@ private class RsLiteralFileReferenceSet(
                 listOfNotNull(item?.containingMod?.getOwnedDirectory())
             }
-            else -> parentDirectoryContext
+            else -> {
+                val rsElement = element as? RsElement ?: return emptyList()
+                val workspaceRoot = rsElement.containingCargoPackage?.contentRoot ?: return emptyList()
+                return toFileSystemItems(workspaceRoot)
+            }
```
Unfortunately, it breaks name resolution and completion in `include` macros and that's why test fails.
For example, the following code is correct
```rust
// -crate_root.txt
/* hello from crate root! */

// -src/sibling.txt
/* hello from src dir! */

// -src/main.rs

use std::fs::File;
use std::io::Read;

fn main() {
    let mut buffer = String::new();
    File::open("crate_root.txt").unwrap().read_to_string(&mut buffer).unwrap();
    println!("{}", buffer);

    let s = include_str!("sibling.txt");
    println!("{}", s);
}
```
and it prints
```
/* hello from crate root! */
/* hello from src dir! */
```
But with these changes, `sibling.txt` cannot be resolved correctly.
So let's use `parentDirectoryContext` for `RsIncludeMacroArgument` as before

Undin reviewed (changes requested):
> I need a nudge. I want to set the directory to element.containingCargoPackage.workspace.contentRoot, i.e. the directory of the containing workspace root. However, this information is provided with a Path, while I need a VirtualFile. I cannot just use the filesystem to resolve the path, because it won't work in tests.
>
> Is there a way to get from Path to VirtualFile so that it works in tests?

As I can see, you've already solved this problem

Kobzol commented:
I actually didn't solve it yet :sweat:
I used `rsElement.containingCargoPackage?.contentRoot`, because it is a `VirtualFile`.
But I want to use `rsElement.containingCargoPackage?.workspace.contentRoot`, but that is a `Path`.
So I managed to get a virtual file, but for the wrong directory.

Undin commented:
> I actually didn't solve it yet 😓
I used rsElement.containingCargoPackage?.contentRoot, because it is a VirtualFile.
But I want to use rsElement.containingCargoPackage?.workspace.contentRoot, but that is a Path.
So I managed to get a virtual file, but for the wrong directory.

Oh, sorry, I've missed this detail. I've tried to provide a proper API for you in #7254

Kobzol commented:
Thank you! My head always spins when I'm trying to remember the difference between workspace, package, crate, project, cargo project etc. :laughing: So I wasn't sure if I missed anything.

Kobzol commented:
I changed it to point to workspace root and fixed the include macro case. I tested it in IDE, it seems to work. To properly test that it resolves to a parent workspace in tests, I guess I would have to create a completion test that would also inherit from `RsWithToolchainTestBase` to get access to a "real" project with a workspace?

Undin commented:
> To properly test that it resolves to a parent workspace in tests, I guess I would have to create a completion test that would also inherit from RsWithToolchainTestBase to get access to a "real" project with a workspace?

@Kobzol I suppose now it's enough to create a custom `org.rust.RustProjectDescriptorBase` with overridden `testCargoProject`

Kobzol commented:
I tried something like this:
```kotlin
@ProjectDescriptor(WithWorkspaceProjectDescriptor::class)
fun `test complete paths in path constructor 1`() = doSingleCompletionByFileTree("""
//- Cargo.toml
    [workspace]
    members = [
        "crate-a",
        "crate-b"
    ]
//- crate-a/src/main.rs
    fn main() {
        std::path::Path::new("crate-a/src/mai/*caret*/");
    }
//- foo.rs
    pub struct Foo;
""", """
    fn main() {
        std::path::Path::new("crate-a/src/main.rs/*caret*/");
    }
""")

object WithWorkspaceProjectDescriptor : WithRustup(DefaultDescriptor) {
    override fun testCargoProject(module: Module, contentRoot: String): CargoWorkspace {
        val crateA = testCargoPackage("$contentRoot/crate-a")
        val crateB = testCargoPackage("$contentRoot/crate-b")

        val packages = listOf(
            crateA,
            crateB
        )
        return CargoWorkspace.deserialize(
            Paths.get("${Urls.newFromIdea(contentRoot).path}/Cargo.toml"),
            CargoWorkspaceData(packages, emptyMap(), emptyMap(), contentRoot), CfgOptions.DEFAULT)
    }
}
```
Is it at least heading into the right direction? :)

Undin commented:
@Kobzol I think it's exactly what we need in the corresponding tests 👍 
Note, since in these "light" tests we don't use actual manifest files at all, we can omit `Cargo.toml` in the tests

Kobzol commented:
It didn't work for me at first, but I managed to get the stdlib + workspace descriptor to work, thanks!

Undin commented on src/test/kotlin/org/rust/lang/core/completion/RsPathCompletionTest.kt:
```diff
@@ -0,0 +1,294 @@
+    """)
+}
+
+object WithWorkspaceProjectDescriptor: RustProjectDescriptorBase() {
```
```suggestion
private object WithWorkspaceProjectDescriptor : RustProjectDescriptorBase() {
```

Undin reviewed (approved):
Thanks!

Kobzol commented:
bors r=Undin

Kobzol commented:
bors retry

Kobzol commented:
Btw @Undin could you please take a look at https://github.com/intellij-rust/intellij-rust/pull/5347? It's been sitting there for almost a year now :(
